source/common/matcher/matcher.h

```diff
@@ -0,0 +1,260 @@
+#pragma once
+
+#include <memory>
+#include <variant>
+
+#include "envoy/config/common/matcher/v3/matcher.pb.h"
+#include "envoy/config/core/v3/extension.pb.h"
+#include "envoy/http/header_map.h"
+#include "envoy/matcher/matcher.h"
+
+#include "common/common/assert.h"
+#include "common/http/header_utility.h"
+
+#include "extensions/common/matcher/matcher.h"
+
+#include "absl/strings/string_view.h"
+#include "absl/types/optional.h"
+
+namespace Envoy {
+
+class MatchWrapper {
```

Comment:
This approach here to support multiple MatchPredicate matchers isn't really feasible once buffering comes into play, as this would buffer each possible MatchPredicate in the tree, regardless of whether they would be used. 

Ideally we'd at the very least want to limit the buffering to only happen once, and ideally we'd only have it happen if the remaining subtree actually requires it.

Reply:
Doesn't the tap matching code already handle all of this? It seems like we can effectively just move that into the filter manager? re: your comment about the buffer filter, at this point the buffer filter is just a tiny wrapper around the buffering functionality in the HCM/FM, so it seems like we can avoid injecting another filter and just rely on the underlying functionality?

Reply:
The main difference is that the match tree allows for multiple MatchPredicates (one per leaf matcher), which means that I needed to add some glue code that would let the MatchingData pass new information (headers/data/etc.) to all of the predicates. Right now each of these match predicates would independently attempt to buffer data in order to do matching on it, even if we already have enough information to exclude a lot of them (e.g. request headers allow us to select some smaller sub tree). 

I'm not sure if the solution here would be to buffer in the HCM (this is a bit tricky with the buffering semantics of the generic body matcher which only buffers the last N bytes required to match the pattern) or to just centralize the data handling in a single object that is shared by all the matchers.

Reply:
>or to just centralize the data handling in a single object that is shared by all the matchers.

I would probably do something like this.